torch/csrc/jit/tensorexpr/kernel.cpp

```diff
@@ -1482,6 +1514,97 @@ BlockPtr TensorExprKernel::bindAllInputs() {
   return block;
 }
 
+void TensorExprKernel::deduceMemoryLayoutPolicy() {
+  // If the tensor is channels-last contiguous, the preferred memory layout
+  // propagation policy is to use channes-last. Otherwise, the preferred policy
+  // is to use contiguous.
+  auto _get_preferred_mem_policy =
+      [](const torch::jit::Value* val,
+         const std::vector<torch::jit::StrideInput>& stride_desc_vec) {
+        if (stride_desc_vec.size() > 0) {
+          // Has symbolic stride information
+          auto cur_stride_desc = stride_desc_vec[0];
+          return (cur_stride_desc ==
+                  torch::jit::StrideInput::TENSOR_CONT_CHANNELS_LAST)
+              ? MemoryLayoutPolicy::kChannelsLastNdContiguous
+              : MemoryLayoutPolicy::kContiguous;
+        }
+
+        const auto& tt = val->type()->expect<TensorType>();
+        // No shape info is present in the graph
+        TORCH_INTERNAL_ASSERT(
+            tt->sizes().concrete_sizes(),
+            buildErrorMessage(
+                "Shapes for " + val->debugName() + " are unknown."));
+        auto sizes = *tt->sizes().concrete_sizes();
+        const std::vector<int64_t> strides = *tt->strides().concrete_sizes();
+        return (c10::is_channels_last_strides_2d(sizes, strides))
+            ? MemoryLayoutPolicy::kChannelsLastNdContiguous
+            : MemoryLayoutPolicy::kContiguous;
+      };
+
+  // Check whether the Value is TensorType
+  auto _is_tensor = [](const jit::Value* el) {
+    if (el->type()->kind() == TypeKind::TensorType) {
+      return true;
+    } else {
+      return false;
+    }
+  };
+
+  std::vector<torch::jit::StrideInput> empty_strides = {};
+
+  // The graph inputs contain Tensor. Othewise, do nothing.
+  auto inputs_contain_tensor =
+      std::any_of(graph_->inputs().begin(), graph_->inputs().end(), _is_tensor);
+  auto inputs_all_channels_last = inputs_contain_tensor &&
+      std::all_of(graph_->inputs().begin(),
+                  graph_->inputs().end(),
+                  [&](const jit::Value* el) {
+                    if (el->type()->kind() != TypeKind::TensorType) {
+                      return true;
+                    }
```

Comment:
If you return `false` directly here, you don't have to check `inputs_contain_tensor` with `_is_tensor` any more, right? Basically, `inputs_all_channels_last` requires all inputs are of tensor type and in channels last format, right?

Reply:
> If you return `false` directly here, you don't have to check `inputs_contain_tensor` with `_is_tensor` any more, right? Basically, `inputs_all_channels_last` requires all inputs are of tensor type and in channels last format, right?

It serves the dynamic shape. The shape information is also part of graph inputs. If the logic returns false directly, then dynamic shape could not support channels-last contiguous tensor.